### PR TITLE
Add new formula for estimating rcf model size

### PR DESCRIPTION
**Pull request by kaituo**
Note: since there are inter-file references, I split a big PR into multiple smaller PRs based on files to save time (1.1 cut off is due at the end of this week). The build will fail due to missing dependencies. I will use that PR just for review. will not merge it. Will have a big PR in the end and merge once after all review PRs get approved. For the complete code, check https://github.com/kaituo/anomaly-detection-1/tree/compactRCF2

### Description
This PR created the formula for estimating compact rcf model size.

Testing done:
1. Compared the actual size and the estimated size. The numbers mostly match.
 
### Issues Resolved
https://github.com/opensearch-project/anomaly-detection/issues/97
 
### Check List
- [ x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by sudiptoguha — APPROVED** [review]
Looks good to me. We should probably have an issue that the formula incorporates samplesize as a variable (now 256). It's ok at the moment - we cannot/should not make that change now. But larger samplesize forests be useful down the road and we would need to change this. But the mechanism of how to change is already here.

**Comment by kaituo** [comment]
> Looks good to me. We should probably have an issue that the formula incorporates samplesize as a variable (now 256). It's ok at the moment - we cannot/should not make that change now. But larger samplesize forests be useful down the road and we would need to change this. But the mechanism of how to change is already here.

Yes, created an issue: https://github.com/opensearch-project/anomaly-detection/issues/146

File: src/main/java/org/opensearch/ad/ml/ModelPartitioner.java [hunk 1]
```diff
@@ -95,6 +97,11 @@ public ModelPartitioner(int rcfNumSamplesInTree, int rcfNumTrees, DiscoveryNodeF
                 .numberOfTrees(rcfNumTrees)
                 .outputAfter(rcfNumSamplesInTree)
                 .parallelExecutionEnabled(false)
+                .compact(true)
+                .precision(Precision.FLOAT_32)
+                .boundingBoxCacheFraction(AnomalyDetectorSettings.REAL_TIME_BOUNDING_BOX_CACHE_RATIO)
+                // same with dimension for opportunistic memory saving
+                .shingleSize(rcfNumFeatures)
```
Comment (ylwu-amzn) [inline src/main/java/org/opensearch/ad/ml/ModelPartitioner.java:104]: Same issue with https://github.com/opensearch-project/anomaly-detection/pull/143#discussion_r677812252

Comment (kaituo) [reply src/main/java/org/opensearch/ad/ml/ModelPartitioner.java:104]: changed

File: src/main/java/org/opensearch/ad/task/ADTaskCacheManager.java [hunk 2]
```diff
@@ -50,19 +53,25 @@
...
+import org.opensearch.ad.model.Entity;
+import org.opensearch.ad.settings.AnomalyDetectorSettings;
 import org.opensearch.cluster.service.ClusterService;
 import org.opensearch.common.settings.Settings;
 import org.opensearch.common.util.set.Sets;
 
 import com.amazon.randomcutforest.RandomCutForest;
+import com.google.common.collect.ImmutableList;
 
 public class ADTaskCacheManager {
```
Comment (kaituo) [inline src/main/java/org/opensearch/ad/task/ADTaskCacheManager.java:67]: @sudiptoguha Ignore most of the diff in this file as they are from rebase.  It's Yaliang's code.

**Review by sudiptoguha — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]

**Pull request closed without merging**